valhalla/baldr/openlr.h

```diff
@@ -161,25 +161,117 @@ struct LocationReferencePoint {
+  BothSidesOfRoad = 3,
+};
+
+// 5.3.2
+// The orientation information (ORI) describes the relationship between the point
+// of interest and the direction of a referenced line. The point may be directed
+// in the same direction as the line, against that direction, both directions,
+// or the direction of the point might be unknown.
+enum Orientation {
+  NoOrientation = 0,
+  FirstLrpTowardsSecond = 1,
+  SecondLrpTowardsFirst = 2,
+  BothDirections = 3,
+};
+
+const uint8_t OPENLR_VERSION = 3;
+
+// The first byte of the OpenLr identifier, page 55
+struct OpenLrStatus {
+  uint8_t version : 3;
+  uint8_t has_attributes : 1;
+  uint8_t ArF0 : 1;
+  uint8_t is_point : 1;
+  uint8_t ArF1 : 1;
+  uint8_t rfu : 1;
+
+  constexpr OpenLrStatus()
+      : version(OPENLR_VERSION), has_attributes(0), ArF0(0), is_point(0), ArF1(0), rfu(0){};
+
+  // Status-byte of a Line segment (page 53)
+  constexpr static OpenLrStatus Line() {
+    OpenLrStatus status;
+    status.version = OPENLR_VERSION;
+    status.has_attributes = 1;
+    status.ArF0 = 0;
+    status.is_point = 0;
+    status.ArF1 = 0;
+    status.rfu = 0;
+    return status;
+  }
+  // Status-byte of a PointAlongLine segment (page 55)
+  constexpr static OpenLrStatus PointAlongLine() {
+    OpenLrStatus status;
+    status.version = OPENLR_VERSION;
+    status.has_attributes = 1;
+    status.ArF0 = 0;
+    status.is_point = 1;
+    status.ArF1 = 0;
+    status.rfu = 0;
+    return status;
+  }
+};
+
+// The first metadata byte after the first coordinate
+struct Attribute5 {
+  uint8_t form_of_way : 3;
+  uint8_t functional_road_class : 3;
+  uint8_t orientation : 2; // Only defined for PointAlongLine
+};
+// The first metadata byte after the non-first coordinate
+struct Attribute6 {
+  uint8_t form_of_way : 3;
+  uint8_t functional_road_class : 3;
+  uint8_t side_of_the_road : 2; // Only defined for PointAlongLine
+};
+
 // Line locations, p.19, section 3.1
 // Only line location with 2 location reference points are supported
-struct LineLocation {
-  LineLocation(const std::string& reference, bool base64_encoded = false) {
+struct OpenLr {
+  OpenLr(const std::string& reference, bool base64_encoded = false) {
     const std::string& decoded = base64_encoded ? decode64(reference) : reference;
-    //  Line location data size: 16 + (n-2)*7 + [0/1/2] bytes
     const size_t size = decoded.size();
-    if (size < 16) {
-      throw std::invalid_argument("OpenLR reference is too small reference=" + reference +
-                                  "size=" + std::to_string(decoded.size()));
-    }
 
-    const auto raw = reinterpret_cast<const unsigned char*>(decoded.data());
+    const auto raw = reinterpret_cast<const uint8_t*>(decoded.data());
     std::size_t index = 0;
 
     // Status, version 3, has attributes, ArF 'Circle or no area location'
-    auto status = raw[index++] & 0x7f;
-    if (status != 0x0b)
+    const OpenLrStatus status = *reinterpret_cast<const OpenLrStatus*>(&raw[index++]);
```

Comment:
you dont really need to make a copy of this, you can just as well have a const reference to it. just need to add the `&` to the declaration